infrastructure/src/main/java/org/corfudb/infrastructure/orchestrator/workflows/AddNodeWorkflow.java

```diff
@@ -162,54 +163,71 @@ private void stateTransfer(String endpoint, CorfuRuntime runtime,
-     * Copies the split segment to the new node, if it
-     * is the new node also participates as a logging unit.
+     * The new server is caught up with all data.
+     * This server is then added to all the segments to mark it open to all reads and writes.
      */
-    protected class StateTransfer extends Action {
+    protected class RestoreRedundancy extends Action {
+        @Nonnull
         @Override
         public String getName() {
-            return "StateTransfer";
+            return "RestoreRedundancy";
         }
 
         @Override
         public void impl(@Nonnull CorfuRuntime runtime) throws Exception {
-            // Transfer the replicated segment to the new node
-            stateTransfer(request.getEndpoint(), runtime, newLayout.getSegment(0));
-        }
-    }
-
-    /**
-     * Merges the fragmented segment if the AddNodeToLayout action caused any
-     * segments to split
-     */
-    protected class MergeSegments extends Action {
-        @Override
-        public String getName() {
-            return "MergeSegments";
-        }
+            runtime.invalidateLayout();
+            newLayout = runtime.getLayoutView().getLayout();
+
+            // A newly added node can be marked as unresponsive by the fault detector by the
+            // time this action is executed. There are 2 cases following this:
+
+            // Case 1. The node remains unresponsive.
+            //      State transfer fails.
+            // Case 2. The node is marked responsive again.
+            //      In this case, the node was removed from all segments and was wiped clean.
+            //      So either the healing workflow or the add node workflow will attempt
+            //      to catchup the new node.
+            if (newLayout.getAllActiveServers().contains(request.endpoint)) {
+                // Transfer only till the second last segment as the last segment is unbounded.
+                // The new server is already a part of the last segment. This is based on an
+                // assumption that the newly added node is not removed from the layout.
+                for (int i = 0; i < newLayout.getSegments().size() - 1; i++) {
+                    stateTransfer(Collections.singleton(request.getEndpoint()),
+                            runtime,
+                            newLayout.getSegments().get(i));
+                }
 
-        @Override
-        public void impl(@Nonnull CorfuRuntime runtime) throws Exception {
-            runtime.getLayoutManagementView().mergeSegments(newLayout);
+                final int stripeIndex = 0;
+                runtime.getLayoutManagementView()
+                        .addLogUnitReplica(
+                                new Layout(newLayout), request.getEndpoint(), stripeIndex);
+                runtime.invalidateLayout();
+                newLayout = runtime.getLayoutView().getLayout();
+            } else {
+                throw new RuntimeException("RestoreRedundancy: "
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Define and throw a dedicated exception instead of using a generic one. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=squid%3AS00112)